client/engine/messageservice/p2p-message-service/service.go

```diff
@@ -0,0 +1,196 @@
+// We ignore peers that are ourselves
+func (ms *P2PMessageService) AddPeers(peers []PeerInfo) {
+
+	for _, p := range peers {
+		// Ignore ourselves
+		if p.Address == ms.me {
+			continue
+		}
+		multi, _ := multiaddr.NewMultiaddr(fmt.Sprintf("/ip4/%s/tcp/%d/p2p/%s", p.IpAddress, p.Port, p.Id))
+
+		// Extract the peer ID from the multiaddr.
+		info, _ := peer.AddrInfoFromP2pAddr(multi)
+		// Add the destination's peer multiaddress in the peerstore.
+		// This will be used during connection and stream creation by libp2p.
+		ms.p2pHost.Peerstore().AddAddrs(info.ID, info.Addrs, peerstore.PermanentAddrTTL)
+
+		ms.peers.Store(p.Address.String(), info.ID)
+	}
+}
+
+// NewMessageService returns a running P2PMessageService listening on the given ip and port
+func NewMessageService(ip string, port int, pk []byte) *P2PMessageService {
+	// We generate a random key using the hash of the pk as a seed
+	// This should mean that the message key is deterministic
+	// TODO: Ideally we would use the pk directly, but I haven't figured out of this is possible with lib p2p
+	hash := sha256.Sum256(pk)
+	seed := big.NewInt(0).SetBytes(hash[:]).Int64()
+	messageKey, _, err := p2pcrypto.GenerateECDSAKeyPair(rand.New(rand.NewSource(seed)))
+	if err != nil {
+		panic(err)
+	}
+	options := []libp2p.Option{libp2p.Identity(messageKey),
+		libp2p.ListenAddrStrings(fmt.Sprintf("/ip4/%s/tcp/%d", ip, port)),
+		libp2p.DefaultTransports,
+		libp2p.NoSecurity,
+		libp2p.DefaultMuxers,
+	}
+	host, err := libp2p.New(options...)
+	if err != nil {
+		panic(err)
+	}
+
+	safePeers := safesync.Map[peer.ID]{}
+	h := &P2PMessageService{
+		out:     make(chan protocols.Message, BUFFER_SIZE),
+		peers:   &safePeers,
+		p2pHost: host,
+		quit:    make(chan struct{}),
+		key:     messageKey,
+		me:      crypto.GetAddressFromSecretKeyBytes(pk),
+	}
+
+	h.p2pHost.SetStreamHandler(MESSAGE_ADDRESS, func(stream network.Stream) {
+
+		select {
+		case <-h.quit:
+			stream.Close()
+			return
+		default:
+
+			reader := bufio.NewReader(stream)
+			// Create a buffer stream for non blocking read and write.
+			raw, err := reader.ReadString(DELIMITER)
+
+			// An EOF means the stream has been closed by the other side.
+			if errors.Is(err, io.EOF) {
+				stream.Close()
+				return
+			}
+			h.checkError(err)
+			m, err := protocols.DeserializeMessage(raw)
+
+			h.checkError(err)
+			h.out <- m
+			stream.Close()
+		}
+
+	})
+
+	return h
+
+}
+
+// Send sends messages to other participants
+// It blocks until the message is sent.
+// It will retry establishing a stream NUM_CONNECT_ATTEMPTS times before giving up
+func (ms *P2PMessageService) Send(msg protocols.Message) {
+	raw, err := msg.Serialize()
+	ms.checkError(err)
+
+	id, ok := ms.peers.Load(msg.To.String())
+	if !ok {
+		panic(fmt.Errorf("could not load peer %s", msg.To.String()))
+	}
+
+	for i := 0; i < NUM_CONNECT_ATTEMPTS; i++ {
+		s, err := ms.p2pHost.NewStream(context.Background(), id, MESSAGE_ADDRESS)
```

Comment:
Similar on the streamHandler. Are we meant to be "looking up" existing streams rather than constructing new ones?